storage/memory/storage.go

```diff
@@ -0,0 +1,268 @@
+package memory
+
+import (
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/math"
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/onflow/flow-evm-gateway/storage"
+	"github.com/onflow/flow-evm-gateway/storage/errors"
+	"github.com/onflow/flow-go/fvm/evm/types"
+	"math/big"
+	"sync"
+)
+
+const (
+	unknownHeight = math.MaxUint64 - 1
+)
+
+type baseStorage struct {
+	mu sync.RWMutex
+
+	blocksIDs       map[common.Hash]*types.Block
+	blockHeightsIDs map[uint64]common.Hash
+	firstHeight     uint64
+	lastHeight      uint64
+
+	receiptsTxIDs       map[common.Hash]*gethTypes.Receipt
+	receiptBlockIDTxIDs map[common.Hash]common.Hash
+	bloomHeight         map[int64]gethTypes.Bloom
+
+	transactionsIDs map[common.Hash]*gethTypes.Transaction
+}
+
+var store *baseStorage
+
+func baseStorageFactory() *baseStorage {
+	if store == nil {
+		store = &baseStorage{
+			blocksIDs:           make(map[common.Hash]*types.Block),
+			blockHeightsIDs:     make(map[uint64]common.Hash),
+			firstHeight:         unknownHeight,
+			lastHeight:          unknownHeight,
+			receiptsTxIDs:       make(map[common.Hash]*gethTypes.Receipt),
+			receiptBlockIDTxIDs: make(map[common.Hash]common.Hash),
+			bloomHeight:         make(map[int64]gethTypes.Bloom),
+			transactionsIDs:     make(map[common.Hash]*gethTypes.Transaction),
+		}
+	}
+	return store
+}
+
+var _ storage.BlockIndexer = &BlockStorage{}
+
+type BlockStorage struct {
+	base *baseStorage
+}
+
+func NewBlockStorage() *BlockStorage {
+	return &BlockStorage{
+		base: baseStorageFactory(),
+	}
+}
+
+func (s BlockStorage) GetByHeight(height uint64) (*types.Block, error) {
+	s.base.mu.RLock()
+	defer s.base.mu.RUnlock()
+
+	// Check if the requested height is within the known range
+	if height < s.base.firstHeight || height > s.base.lastHeight {
+		return nil, errors.NotFound
+	}
+
+	// Retrieve the block using the blockHeightsIDs map
+	blockID, exists := s.base.blockHeightsIDs[height]
+	if !exists {
+		return nil, errors.NotFound
+	}
+
+	// Retrieve the block using the blocksIDs map
+	block, exists := s.base.blocksIDs[blockID]
+	if !exists {
+		return nil, errors.NotFound
+	}
+
+	return block, nil
+}
+
+func (s BlockStorage) GetByID(ID common.Hash) (*types.Block, error) {
+	s.base.mu.RLock()
+	defer s.base.mu.RUnlock()
+
+	// Retrieve the block using the blocksIDs map
+	block, exists := s.base.blocksIDs[ID]
+	if !exists {
+		return nil, errors.NotFound
+	}
+
+	return block, nil
+}
+
+func (s BlockStorage) Store(block *types.Block) error {
+	s.base.mu.Lock()
+	defer s.base.mu.Unlock()
+
+	ID, err := block.Hash()
+	if err != nil {
+		return fmt.Errorf("block hash error: %w", err)
+	}
```

Comment:
You might switch the order of this to minimize the block time,

Reply:
hmm not sure what you mean, order of which? the check?

Reply:
ohh I see, that makes sense yeah